xwiki-platform-core/xwiki-platform-web/src/main/webapp/resources/uicomponents/suggest/suggestPages.js

```diff
@@ -0,0 +1,90 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+require.config({
+  paths: {
+    'xwiki-selectize': "$xwiki.getSkinFile('uicomponents/suggest/xwiki.selectize.js', true)" +
+      "?v=$escapetool.url($xwiki.version)"
+  }
+});
+
+define('xwiki-suggestPages', ['jquery', 'xwiki-selectize'], function($) {
+  'use strict';
+
+  var pageIcon = $jsontool.serialize($services.icon.getMetaData('page_white'));
+
+  var getSelectizeOptions = function(select) {
+    return {
+      create: true,
+      load: function(text, callback) {
+        loadPages(text).done(function(data) {
+          var pages = [];
+          data.searchResults.forEach(function (element) {
+            var label = element.title;
+            var queryString = "";
+            // Separate spaces with " / " and unescape characters. E.g. A\.B.C\\D => A.B / C\D
+            var hint = element.space.replace(/([^\\])\./g, "$1 / ").replace(/\\(.)/g, "$1");
```

Comment:
This is error-prone and this code shouldn't be aware or care about the reference syntax. Moreover, [the ``space`` field is deprecated](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-server/xwiki-platform-search-solr-server-data/src/main/resources/xwiki/conf/managed-schema#L133). If it weren't deprecated you should have resolved the space reference (as you do below with the document reference). But since it is, you could use the ``spaces`` field which is an array of space names, making it even easier to compute the hint. But, if you do this, the hint will still be made of space names not titles. The Page property picker computes the hint by rendering the title of each parent page. Same does the breadcrumb displayed on top of each page. We should display the "space" titles here also. One option is to make an AJAX request to ``hierarchy_reference.vm`` for each search result (suggestion), but it may be too costly..

Reply:
> I am not able to find the ``spaces`` field

Sorry, I thought the search response matches the data we have in the [Solr index](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-server/xwiki-platform-search-solr-server-data/src/main/resources/xwiki/conf/managed-schema#L133). I was referring to the ``space`` and ``spaces`` fields from the Solr index, but the response is actually the serialization of a list of [``SearchResult``](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-model/src/main/resources/xwiki.rest.model.xsd#L540), which is part of the REST model and generic (used by all query types, [``SOLRSearchSource``](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-rest/src/main/java/org/xwiki/search/solr/internal/rest/SOLRSearchSource.java#L160) being just one of the implementations).

>  add a ``spaces`` field in the ``/wikis/{wikiName}/search`` endpoint

We could but we don't need it, since you can use the [JavaScript API](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/FrontendResources/JavaScriptAPI/#HWorkwithEntityReferences) to resolve the value of the ``space`` field.

Reply:
> We could but we don't need it, since you can use the JavaScript API to resolve the value of the `space` field.

Sure but it doesn't give you the title of the spaces. Should we just display the spaces (or document) name for now? As you said resolving the space title with AJAX calls is very inefficient.

Reply:
I see. I wouldn't put the space titles in the ``spaces`` field anyway. I would use a different field, say ``breadcrumb`` or ``hierarchy`` or ``path``. I'd keep ``spaces`` for the space names. I'm OK to add this to the ``SearchResult`` model.